### ISSUE TITLE
Simplify Animation calculations

### ISSUE THREAD

**Issue by mourner** [Leaflet/Leaflet#3964]
This looks bad: https://github.com/Leaflet/Leaflet/blob/master/src/layer/vector/Renderer.js#L56

We need a really simple way to do animation calculations so that plugin developers could add animation easily.

**Comment by cmulders** [comment]
The calculations become much simpler when the transform-origin is set to the middle, then the scaling does not influence the translation during the transform. This reduces the ```viewhalf``` calculation and the complicated ```topLeftOffset```

Made pull request with the changes

**Comment by mourner** [comment]
Thanks! The problem is though that we also want the animation code between different layers (vectors, tile layers, image overlays, etc.) consistent. It was calculated with 50% origin earlier, but I switched it to `0 0` for every layer because having it different made things even crazier.

**Comment by cmulders** [comment]
Ah I see your change (1e9c864). Hmm, then it becomes indeed more complicated, possibly some utility function to calculate the correct transform-offset for a given offset + scale? Reducing the math for plugin authors?

**Comment by mourner** [comment]
@cmulders yes, exactly. Preferably we'd come up with a universal public method for things like this. I'm going to try rethinking this today with a fresh mind. This is a really brainmelting task — I remember my previous attempts.

**Comment by cmulders** [comment]
I think that using the existing L.Transformation (possibly implementing Homogenous coordinates) consistently is the best option. Keeping track of a 'origin point' in every layer and multiplying this with the transformation matrix when a zoom animation is started. 

I tried to implement something, but I could not get setZoomAround working properly, because the origin of scaling is shifted (and must be done by the transformation). Seriously brainmelting..

**Comment by mourner** [comment]
@cmulders thanks for looking into this! Now you see what I mean by the `brainmelt` label...
